full modern bert support

ryan-mangeno opened the pull request:
Made support for conversion from hf->gguf and execution on llama.cpp after my recent (granite-embd-support)[https://github.com/ggml-org/llama.cpp/pull/15641] which is a modern bert based model, this pr continues off of that and has some tweaks. I have ran cosine similarity tests with this script

```
from sentence_transformers import SentenceTransformer
import numpy as np
import subprocess
import shlex
import os

model_path = os.path.expanduser(
    "~/models/models--answerdotai--ModernBERT-large/snapshots/45bb4654a4d5aaff24dd11d4781fa46d39bf8c13/"
)
lcpp_model = os.path.expanduser("~/models/modern-bert-large.gguf")
lcpp_exe = "/Users/ryanmangeno/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding"

model = SentenceTransformer(model_path)

input_queries = [
    "hello world",
    "tell me a story about a developer and their dog",
    "123sfg this is a r@nd0m t35t",
]


def cosine_similarity(vector_a: np.ndarray, vector_b: np.ndarray) -> float:
    vector_a = np.asarray(vector_a)
    vector_b = np.asarray(vector_b)
    numerator = np.dot(vector_a, vector_b)
    denominator_a = np.linalg.norm(vector_a)
    denominator_b = np.linalg.norm(vector_b)
    if denominator_a == 0 or denominator_b == 0: return 0.0
    cosine_sim = numerator / (denominator_a * denominator_b)
    return cosine_sim


for query in input_queries:
    print("### BASELINE ###")
    embedding = model.encode([query])
    print("Embedding shape:", embedding.shape)
    print("Embedding vector:", embedding[:, :8])

    print("### llama.cpp ###")
    cmd = f"{lcpp_exe} -m {lcpp_model} -p \"{query}\" --temp 0 --embd-normalize -1 --pooling mean"
    print(f"llama.cpp command: {cmd}")
    proc = subprocess.Popen(
        shlex.split(cmd),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    out, _ = proc.communicate()
    vals = out.decode("utf-8").split(":")[-1]
    vals = [
        float(v) for v in vals.split()
        if v.strip()
    ]
    lcpp_emb = np.array(vals)
    print("llama.cpp Embedding shape:", lcpp_emb.shape)
    print("llama.cpp Embedding vector:", lcpp_emb[:8])
    print()
    cos_sim = cosine_similarity(embedding, lcpp_emb)
    print(f"COSINE SIMILARITY: {cos_sim}")
    print("--------------------------------")
    print()
```

with the following results
### BASELINE ###
Embedding shape: (1, 1024)
Embedding vector: [[ 0.659244    0.39849958  0.2302168   0.6192862  -0.62407815  0.0042014
   0.14638135  0.2541136 ]]
### llama.cpp ###
llama.cpp command: /Users/ryanmangeno/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding -m /Users/ryanmangeno/models/modern-bert-large.gguf -p "hello world" --temp 0 --embd-normalize -1 --pooling mean
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
llama.cpp Embedding shape: (1024,)
llama.cpp Embedding vector: [ 0.679083  0.394328  0.235899  0.62807  -0.630304 -0.004468  0.141795
  0.248705]

COSINE SIMILARITY: [0.99971951]
--------------------------------

### BASELINE ###
Embedding shape: (1, 1024)
Embedding vector: [[ 0.23057191  0.12633912 -0.00238159 -0.08394846  0.19630949  0.03715154
   0.0040304   0.63173795]]
### llama.cpp ###
llama.cpp command: /Users/ryanmangeno/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding -m /Users/ryanmangeno/models/modern-bert-large.gguf -p "tell me a story about a developer and their dog" --temp 0 --embd-normalize -1 --pooling mean
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
llama.cpp Embedding shape: (1024,)
llama.cpp Embedding vector: [ 0.230692  0.131872 -0.007958 -0.065828  0.282647  0.056364 -0.025206
  0.672672]

COSINE SIMILARITY: [0.9994365]
--------------------------------

### BASELINE ###
Embedding shape: (1, 1024)
Embedding vector: [[ 0.15972608  0.52267325 -0.05636618  0.40699816  0.6401572   0.49469572
  -0.4336093   0.3909793 ]]
### llama.cpp ###
llama.cpp command: /Users/ryanmangeno/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding -m /Users/ryanmangeno/models/modern-bert-large.gguf -p "123sfg this is a r@nd0m t35t" --temp 0 --embd-normalize -1 --pooling mean
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
llama.cpp Embedding shape: (1024,)
llama.cpp Embedding vector: [ 0.177373  0.495097 -0.114586  0.46121   0.635596  0.548017 -0.400412
  0.430722]

COSINE SIMILARITY: [0.99780866]
--------------------------------

running the same tests on the granite-embd-small gives same results as before
@gabe-l-hart

ryan-mangeno committed:
full modern bert support

CISC commented:
Can you test f.ex. `Alibaba-NLP/gte-reranker-modernbert-base`?

ryan-mangeno commented:
> Can you test f.ex. `Alibaba-NLP/gte-reranker-modernbert-base`?

yes will do that right now :))

CISC commented:
> > Can you test f.ex. `Alibaba-NLP/gte-reranker-modernbert-base`?
> 
> yes will do that right now :))

I meant specifically with reranking, as that's when `cls_norm` will be used.

ryan-mangeno commented:
> > > Can you test f.ex. `Alibaba-NLP/gte-reranker-modernbert-base`?
> > 
> > 
> > yes will do that right now :))
> 
> I meant specifically with reranking, as that's when `cls_norm` will be used.

I see a mismatch in the ranking logits so Im going to keep working at this

ryan-mangeno committed:
added gelu op in rank pooling for modern bert

ryan-mangeno commented:
> > matching is much better now, but unsure if this is this much diff is acceptable
> 
> Hmmm, unsure too, would be nice to figure out where the deviation starts.


still working on this, still has been a bit tough to close gap, will keep updating

ryan-mangeno commented:
> > matching is much better now, but unsure if this is this much diff is acceptable
> 
> Hmmm, unsure too, would be nice to figure out where the deviation starts.

found some things to note, I made a script to compare embeddings before pooling is applied as I am currently trying to find when they diverge

here is the script:
```
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import numpy as np
import subprocess
import shlex
import os
import torch
import re

os.environ["TOKENIZERS_PARALLELISM"] = "false"

model_path = os.path.expanduser("~/models/models--Alibaba-NLP--gte-reranker-modernbert-base/snapshots/f7481e6055501a30fb19d090657df9ec1f79ab2c/")
lcpp_model = os.path.expanduser("~/models/modern-bert-base.gguf")
lcpp_exe = "/Users/ryanmangeno/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding"

tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
model = AutoModelForSequenceClassification.from_pretrained(model_path, trust_remote_code=True)
model.eval()

input_queries = [
    "hello world",
    "tell me a story about a developer and their dog",
    "123sfg this is a r@nd0m t35t",
]

def get_lcpp_unpooled_vectors(query):
    cmd = f"{lcpp_exe} -m {lcpp_model} -p \"{query}\" --pooling none --embd-normalize -1 -ngl 0"
    proc = subprocess.run(shlex.split(cmd), capture_output=True, text=True)

    clean_stdout = proc.stdout.replace("...", " ")

    blocks = re.split(r'embedding \d+:', clean_stdout)[1:]

    vectors = []
    for b in blocks:
        num_strings = re.findall(r"[-+]?\d*\.\d+(?:[eE][-+]?\d+)?", b)
        vec = np.array([float(x) for x in num_strings])

        if len(vec) > 0:
            vectors.append(vec)

    return vectors

for query in input_queries:
    print(f"### EVALUATING QUERY: '{query}' ###")

    inputs = tokenizer(query, return_tensors="pt")
    tokens = tokenizer.convert_ids_to_tokens(inputs['input_ids'][0])

    with torch.no_grad():
        outputs = model.model(**inputs)
        # captures last hidden states and pass then through final norm, these are embeddings just before pooling is applied
        hf_states = model.model.final_norm(outputs.last_hidden_state)[0].cpu().numpy()

    lcpp_vecs = get_lcpp_unpooled_vectors(query)

    print(f"{'Index':<5} | {'Token':<18} | {'Cosine Sim':<12} | {'Status'}")
    print("-" * 55)

    num_to_compare = min(len(tokens), len(lcpp_vecs))

    if num_to_compare == 0:
        print("FAILED TO CAPTURE LLAMA.CPP VECTORS. Check your binary path.")

    for i in range(num_to_compare):
        v_hf = hf_states[i]
        v_lcpp = lcpp_vecs[i]

        hf_slice = np.concatenate([v_hf[:3], v_hf[-3:]])
        lcpp_slice = np.concatenate([v_lcpp[:3], v_lcpp[-3:]])

        dot = np.dot(hf_slice, lcpp_slice)
        sim = dot / (np.linalg.norm(hf_slice) * np.linalg.norm(lcpp_slice))

        status = "MATCH" if sim > 0.99 else "DRIFT"
        print(f"{i:<5} | {tokens[i]:<18} | {sim:<12.6f} | {status}")

    if len(tokens) != len(lcpp_vecs):
        print(f"WARN: Token count mismatch! HF={len(tokens)}, LCPP={len(lcpp_vecs)}")

    print("-" * 55 + "\n")
```

resulting in ...

```
### EVALUATING QUERY: 'hello world' ###
Index | Token              | Cosine Sim   | Status
-------------------------------------------------------
0     | [CLS]              | 0.999257     | MATCH
1     | hello              | 0.999263     | MATCH
2     | Ġworld             | 0.999257     | MATCH
3     | [SEP]              | 0.999258     | MATCH
-------------------------------------------------------

### EVALUATING QUERY: 'tell me a story about a developer and their dog' ###
Index | Token              | Cosine Sim   | Status
-------------------------------------------------------
0     | [CLS]              | 0.999162     | MATCH
1     | tell               | 0.999078     | MATCH
2     | Ġme                | 0.999142     | MATCH
3     | Ġa                 | 0.999121     | MATCH
4     | Ġstory             | 0.999135     | MATCH
5     | Ġabout             | 0.999055     | MATCH
6     | Ġa                 | 0.999210     | MATCH
7     | Ġdeveloper         | 0.999155     | MATCH
8     | Ġand               | 0.999138     | MATCH
9     | Ġtheir             | 0.999166     | MATCH
10    | Ġdog               | 0.999227     | MATCH
11    | [SEP]              | 0.999244     | MATCH
-------------------------------------------------------

### EVALUATING QUERY: '123sfg this is a r@nd0m t35t' ###
Index | Token              | Cosine Sim   | Status
-------------------------------------------------------
0     | [CLS]              | 0.998803     | MATCH
1     | 123                | 0.998685     | MATCH
2     | sf                 | 0.998793     | MATCH
3     | g                  | 0.998738     | MATCH
4     | Ġthis              | 0.998776     | MATCH
5     | Ġis                | 0.998652     | MATCH
6     | Ġa                 | 0.998863     | MATCH
7     | Ġr                 | 0.998889     | MATCH
8     | @                  | 0.998980     | MATCH
9     | nd                 | 0.998906     | MATCH
10    | 0                  | 0.999116     | MATCH
11    | m                  | 0.998838     | MATCH
12    | Ġt                 | 0.999021     | MATCH
13    | 35                 | 0.999361     | MATCH
14    | t                  | 0.999454     | MATCH
15    | [SEP]              | 0.998584     | MATCH
-------------------------------------------------------
```

ryan-mangeno committed:
still working on stuff, added mean calculation before classifier head

CISC commented:
> the only addition that the modernbert-gte-reranker makes it seems is that after, it applies pooling using a [prediction head](https://github.com/huggingface/transformers/blob/main/src/transformers/models/modernbert/modular_modernbert.py#L1144-L1154) which seems to do cls -> dense -> act -> norm, further, they [apply mean pooling before ranking](https://github.com/huggingface/transformers/blob/main/src/transformers/models/modernbert/modular_modernbert.py#L1404-1411) as in the config.json, it is set to mean, and does mean -> head -> classifier, I believe that is what I have exactly done when the rank pooling is build but im still seeing variance, im unsure if my my logic correct, could you give me some insight? thank you!!

None ATM unfortunately, sounds like you are on the right track though.

ryan-mangeno committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

gabe-l-hart commented:
@ryan-mangeno Taking [this comment](https://github.com/ggml-org/llama.cpp/pull/17910/changes#r2684278802) over here to avoid flooding that old PR with more notifications.

>alright, thanks for letting me know though @gabe-l-hart ! I just got back to school but I have been doing quite a bit of debugging between transformers and llama.cpp in terms of finding where they diverge, if its useful ... I have been inspecting the pooling process much closer and using the llama-eval-callback tool, but there is no pooling flag to be passed, would it make sense to add support for this / expose it. In my case I can hardcode it in llama-embedding.cpp and view the tensors made during build_pooling.

I'm honestly not that familiar with Modern Bert, but I bet @hansolosan might have some thoughts/opinions (he's the lead behind the Granite Embedding models).

ryan-mangeno commented:
hey, just an update ... still working on this but have been very busy as of late, hope to have this fixed soon

ryan-mangeno commented:
I think I did it !!!!! @CISC

ryan-mangeno commented:
the fix is also embarrassingly short (literally one word after countless hours of debugging, testing, and headaches, but also alot of learning!!) haha, I overlooked it because it was actually a flaw in my original implementation regarding the granite-embd support, specifically with setting the swa pattern, I completely forget to set dense first as true, which then made the rope freq incorrect for the first layer! a theta of 10,000 vs the correct theta of 160,000, here is the rank testing script I used as well :) 

```python
import torch
import numpy as np
import subprocess
import shlex
import os
from transformers import AutoTokenizer, AutoModelForSequenceClassification

os.environ["TOKENIZERS_PARALLELISM"] = "false"

model_path = os.path.expanduser("~/models/models--Alibaba-NLP--gte-reranker-modernbert-base/snapshots/f7481e6055501a30fb19d090657df9ec1f79ab2c/")
lcpp_model = os.path.expanduser("~/models/modern-bert-rerank.gguf")
lcpp_exe = os.path.expanduser("~/Projects/gits/llama-fix/llama.cpp/build/bin/llama-embedding")

pairs = [
    ["hello", "world"],
    ["what is the capital of China?", "Beijing"],
    ["how to implement quick sort in python?", "Introduction of quick sort"],
    ["Define photosynthesis in simple terms.", "Photosynthesis is the process where plants use sunlight to make food."],
    ["Who wrote the play Romeo and Juliet?", "William Shakespeare is the author of Romeo and Juliet."],
    ["What is the boiling point of water?", "Pure water boils at 100 degrees Celsius at sea level."],
    
    ["best way to cook salmon", "Salmon is a fatty fish rich in omega-3."],
    ["How do I fix a flat tire?", "You need a spare tire and a jack to change a wheel."],
    ["What are the symptoms of the flu?", "Common symptoms include fever, cough, and sore throat."],
    ["latest stock market trends 2026", "Investors are looking at renewable energy and AI sectors this year."],
    ["How to learn C++ for beginners", "Python is often recommended as a first programming language."],
    
    ["how to implement quick sort in python?", "The weather is nice today"],
    ["What is the capital of France?", "The Eiffel Tower is located in Paris, which is in France."],
    ["How to bake a chocolate cake", "To build a computer, you first need a motherboard and a CPU."],
    ["What is the distance to the moon?", "The moon has a profound effect on Earth's ocean tides."],
    ["Benefits of regular exercise", "The price of gold has increased significantly over the last decade."]
]


print(f"Loading HF Model from: {model_path}")
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
model = AutoModelForSequenceClassification.from_pretrained(
    model_path, 
    trust_remote_code=True,
    dtype=torch.float32,
)
model.eval()

print("\n=== STARTING COMPARISON ===\n")

for query, doc in pairs:
    #  hf Score
    with torch.no_grad():
        inputs = tokenizer([[query, doc]], padding=True, truncation=True, return_tensors='pt', max_length=512)
        input_ids = inputs['input_ids'][0].tolist()
        tokens = tokenizer.convert_ids_to_tokens(input_ids)
        
        print(f"HF Prompt")
        print(f"Query: {query}")
        print(f"Doc:   {doc}")
        print(f"Tokens: {' '.join(tokens)}")
        logits = model(**inputs).logits.view(-1).float()
        hf_score = logits.item()
    
    #  llama.cpp score
    lcpp_prompt = f"{query}[SEP]{doc}"
    
    cmd = f'{lcpp_exe} -m {lcpp_model} -p "{lcpp_prompt}" --embd-separator "|||" --pooling rank -fa on -ngl 0 --embd-normalize -1'
    
    proc = subprocess.Popen(
        shlex.split(cmd),
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
    )
    out, _ = proc.communicate()
    out_str = out.decode("utf-8")

    lcpp_score = None
    try:
        for line in out_str.splitlines():
            if "rerank score 0:" in line:
                lcpp_score = float(line.split(":")[-1].strip())
                break
        
        if lcpp_score is not None:
            diff = abs(lcpp_score - hf_score)
            match_status = "MATCH" if diff < 0.05 else "MISMATCH"
            
            w = 20
            print(f"{'Pair:':<{w}} {query} | {doc}")
            print(f"{'HF Score:':<{w}} {hf_score:.6f}")
            print(f"{'Llama.cpp Score:':<{w}} {lcpp_score:.6f}")
            print(f"{'Difference:':<{w}} {diff:.6f} [{match_status}]")
            print("-" * 60)
        else:
            print(f"FAILED TO PARSE. Raw Output:\n{out_str}")

    except Exception as e:
        print(f"Error processing output: {e}")

print("\n=== COMPARISON COMPLETE ===\n")
```

ryan-mangeno committed:
first layer is dense, as per modern bert research paper

ryan-mangeno committed:
Merge branch 'master' into full-modern-bert-support

CISC commented on src/llama-graph.cpp:
```diff
@@ -185,7 +185,7 @@ bool llm_graph_input_out_ids::can_reuse(const llm_graph_params & params) {
 
 void llm_graph_input_mean::set_input(const llama_ubatch * ubatch) {
-    if (cparams.embeddings && cparams.pooling_type == LLAMA_POOLING_TYPE_MEAN) {
+    if ((cparams.embeddings && cparams.pooling_type == LLAMA_POOLING_TYPE_MEAN) || true /*temporary will not be included on final pr, just to allow set input to run*/ ) {
```
```suggestion
    if (cparams.embeddings && cparams.pooling_type == LLAMA_POOLING_TYPE_MEAN) {
```

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -11003,12 +11003,16 @@ def set_gguf_parameters(self):
         if name.startswith("model."):
             name = name[6:]
+            
+        if self.cls_out_labels:
```
```suggestion

        if self.cls_out_labels:
```

CISC commented:
> assuming all is well, is there something that you all at llama.cpp have been looking to implement wether thats another model and/or feature? I would be glad to look into some more stuff !

You may want to check out the [good first issue](https://github.com/ggml-org/llama.cpp/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22) list.

ryan-mangeno committed:
Update src/llama-graph.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ryan-mangeno committed:
fixed set input for mean pooling to check if pooling type is ranking since modern bert does mean & rank

ryan-mangeno committed:
Update src/llama-graph.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ryan-mangeno committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>
